File: 518-coin-change-ii/coin-change-ii.cpp
class Solution {
public:
    unordered_map<string, int> catche;

    int dfs(int i, int a, int amount, vector<int>& coins){
        if (a > amount) return 0;
        if (a == amount) return 1;
        if (i == coins.size()) return 0;
        string key = to_string(i) + "_" + to_string(a);
        if (catche.contains(key)) return catche[key];

        catche[key] = dfs(i + 1, a, amount, coins) + dfs(i, a + coins[i], amount, coins);
        return catche[key];
    }

    int change(int amount, vector<int>& coins) {
        return dfs(0,0, amount, coins);

    }
};

File: 518-coin-change-ii/coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        df = [0 for _ in range(amount + 1) ]
        df[0] = 1
        for i in range(len(coins) - 1, -1, -1):
            temp = [0 for _ in range(amount + 1)]
            temp[0] = 1
            for j in range(1, amount + 1):
                temp[j] = df[j]
                if j - coins[i] >=0:
                    temp[j] += temp[j - coins[i]]
            df = temp
        return df[amount] 

        # df = [[0 for _ in range(len(coins) + 1)] for _ in range(amount + 1)]
        # df[0] = [1 for _ in range(len(coins) + 1)]

        # for i in range(1, amount + 1):
        #     for j in range(len(coins) - 1, -1, -1):
        #         df[i][j] = df[i][j+1]
        #         if i - coins[j]>=0:
        #             df[i][j] += df[i - coins[j]][j]


        # return df[amount][0]

        # catche = {}

        # def dfs(i, a):
        #     if a == amount: return 1
        #     if a > amount: return 0
        #     if i == len(coins): return 0
        #     if (i, a) in catche: return catche[(i, a)]

        #     catche[(i, a)] = dfs(i+1, a) + dfs(i, a + coins[i])
        #     return catche[(i, a)]
        # return dfs(0,0)